### PR TITLE
Raise pretty error message when connection fails

### LINKED ISSUE #81: When there is an issue connecting, try common username/passwords and suggest them
For example, different distros of Linux have different default usernames and passwords. Would be cool to automatically try those and suggest one if it works

### PR DESCRIPTION
**Pull request by jwoertink**
closes #81 (though, this is an alternative to doing that)

Prior to this, you'd get a `DB::ConnectionRefused error` as well as some `PQ` related errors. Now you get a nice error message that helps to explain what went wrong, and possible ways to fix it.

### CODE REVIEW

File: spec/connection_spec.cr [hunk 1]
```diff
@@ -0,0 +1,11 @@
+require "./spec_helper"
+
+describe Avram::Connection do
+  it "displays a helpful error when failing to connect" do
+    conn = Avram::Connection.new("postgres://root:root@localhost:5432/tacoman")
+    message = Regex.new("Failed to connect to databse 'tacoman' with username root.")
```
Comment (paulcsmith) [inline spec/connection_spec.cr]: I believe you can pass the string and it will do a substring match. I'm not 100% sure though but worth a try

File: src/avram/connection.cr [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  def initialize(@connection_string : String)
+  end
+
+  def try_connection!
+    begin
+      DB.open(@connection_string)
+    rescue DB::ConnectionRefused
+      raise ConnectionError.new(connection_uri)
+    end
```
Comment (paulcsmith) [inline src/avram/connection.cr]: You can remove the begin and end and it will still work. This is a handy little shortcut for rescuing errors

File: src/avram/errors.cr [hunk 3]
```diff
@@ -59,4 +59,22 @@ module Avram
...
+
+  class ConnectionError < AvramError
+    def initialize(connection_details : URI)
+      error = String.build do |message|
+        message << "Failed to connect to databse '#{connection_details.path.try(&.[1..-1])}' with username #{connection_details.user}.\n"
+        message << "You can try these steps to help:"
+        message << '\n'
+        message << '\n'
+        message << "  ▸ Ensure you have the correct settings in config/database.cr\n"
+        message << "  ▸ Does the database exist? If not, be sure to create it.\n"
```
Comment (paulcsmith) [inline src/avram/errors.cr]: Might be worth including the command `lucky db.create`

File: src/avram/errors.cr [hunk 4]
```diff
@@ -59,4 +59,22 @@ module Avram
...
+        message << "  ▸ Check that you have access to connect to #{connection_details.host} on port #{connection_details.port}\n"
+        if connection_details.password.blank?
+          message << "  ▸ You didn't supply a password, did you mean to?\n"
```
Comment (paulcsmith) [inline src/avram/errors.cr]: Nice 👍

File: src/avram/errors.cr [hunk 5]
```diff
@@ -59,4 +59,22 @@ module Avram
...
   end
+
+  class ConnectionError < AvramError
+    def initialize(connection_details : URI)
+      error = String.build do |message|
+        message << "Failed to connect to databse '#{connection_details.path.try(&.[1..-1])}' with username #{connection_details.user}.\n"
+        message << "You can try these steps to help:"
+        message << '\n'
+        message << '\n'
+        message << "  ▸ Ensure you have the correct settings in config/database.cr\n"
```
Comment (paulcsmith) [inline src/avram/errors.cr]: Nice. We can add some comments here in the generated code to help people find a suitable username/password for command Postgres installations

File: src/avram/errors.cr [hunk 6]
```diff
@@ -59,4 +59,22 @@ module Avram
...
     end
   end
+
+  class ConnectionError < AvramError
+    def initialize(connection_details : URI)
+      error = String.build do |message|
+        message << "Failed to connect to databse '#{connection_details.path.try(&.[1..-1])}' with username #{connection_details.user}.\n"
+        message << "You can try these steps to help:"
+        message << '\n'
+        message << '\n'
```
Comment (paulcsmith) [inline src/avram/errors.cr]: I think these should be double quotes but I am not sure.

File: spec/connection_spec.cr (hunk 1 above)
Comment (jwoertink) [reply spec/connection_spec.cr]: I tried that first and it fails. Seems the string does exact match?

File: src/avram/errors.cr (hunk 6 above)
Comment (jwoertink) [reply src/avram/errors.cr]: They should be single because the carriage return is a single character. It's actually more performant to do it this way though negligible when talking about this small of code...

Comment (paulcsmith) [reply src/avram/errors.cr]: TIL!

**Comment by jwoertink** [comment]
Screenshot of what the error looks like:

Screen Shot 2019-04-26 at 10 24 06 AM

File: src/avram/errors.cr (hunk 3 above)
Comment (paulcsmith) [reply src/avram/errors.cr]: Also I think this can be shortened. `Be sure the database exists (lucky db.create)`

File: src/avram/errors.cr (hunk 5 above)
Comment (paulcsmith) [reply src/avram/errors.cr]: People don't read (myself included) so maybe we can shorten this to `Check connection settings in 'config/databse.cr'`

File: src/avram/errors.cr [hunk 7]
```diff
@@ -59,4 +59,22 @@ module Avram
       super message
     end
   end
+
+  class ConnectionError < AvramError
+    def initialize(connection_details : URI)
+      error = String.build do |message|
+        message << "Failed to connect to databse '#{connection_details.path.try(&.[1..-1])}' with username #{connection_details.user}.\n"
```
Comment (paulcsmith) [inline src/avram/errors.cr]: Let's wrap the username in `'` as well (like with the database name

File: src/avram/errors.cr [hunk 8]
```diff
@@ -59,4 +59,22 @@ module Avram
...
+        message << "You can try these steps to help:"
```
Comment (paulcsmith) [inline src/avram/errors.cr]: In most places we have this as `Try this...`

File: spec/connection_spec.cr [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    expect_raises(Avram::ConnectionError, message) do
+      conn.try_connection!
+    end
+  end
```
Comment (paulcsmith) [inline spec/connection_spec.cr]: Could you add a spec for one without password to check that it displays the password hint?

**Comment by paulcsmith** [comment]
Also, I have a feeling you like Tacos...

**Comment by paulcsmith** [comment]
If it isn't too much to ask, could you add some common usernames/passwords to https://github.com/luckyframework/lucky_cli/blob/2cffd3fdd9e288ae20cd06cba8cd82bb12ff7e8c/src/web_app_skeleton/config/database.cr.ecr#L7 ?
